hrms/overrides/employee_advance.py

```diff
@@ -0,0 +1,201 @@
+            FROM `tabExpense Claim Advance` eca, `tabExpense Claim` ec
+            WHERE
+                eca.employee_advance = %s
+                AND ec.name = eca.parent
+                AND ec.docstatus=1
+                AND eca.allocated_amount > 0
+        """,
+                self.name,
+            )[0][0]
+            or 0
+        )
+
+        frappe.db.set_value("Employee Advance", self.name, "claimed_amount", flt(claimed_amount))
+        self.reload()
+        self.set_status(update=True)
+
+    def set_status(self, update=False):
+        status = self.status
+        if self.sanctioned_amount == self.paid_amount:
+            self.status = "Paid"
+
+        if update:
+            self.db_set("status", status)
+        else:
+            self.status = status
+
+    def after_insert(self):
+        if self.status == "Draft":
+            user_permission_doc = frappe.get_doc(
+                {
+                    "doctype": "User Permission",
+                    "user": self.approver_1,
+                    "allow": "Employee Advance",
+                    "for_value": self.name,
+                    "apply_to_all_doctypes": 1,
+                }
+            )
+            doc = user_permission_doc.insert()
+
+
+@frappe.whitelist()
+def user_has_role(user, role):
+    role_present = frappe.db.get_list(
+        "Has Role", filters={"parent": user, "role": role}, ignore_permissions=True
+    )
+    if role_present:
+        return True
+    else:
+        return False
+
+
+@frappe.whitelist()
+def next_state(doc):
+    import json
+
+    if isinstance(doc, str):
+        doc = frappe._dict(json.loads(doc))
+
+    employee_advance_doc = frappe.get_doc("Employee Advance", doc.name)
+
+    if employee_advance_doc.status == "Draft":
+        if frappe.session.user != doc.owner:
+            frappe.throw(_("Only Owner Approve in draft state"))
+        CustomTask._create_link_doc_task(
+            "",
+            "Employee Advance",
+            doc.name,
+            TaskType.APPROVE_EMPLOYEE_ADVANCE,
+            "Medium",
+            "",
+            [f"{doc.approver_1}"],
+        )
+        employee_advance_doc.status = "Pending Approval"
+    elif employee_advance_doc.status == "Pending Approval":
+        if not user_has_role(frappe.session.user, RoleEnum.EXPENSE_APPROVER1_ROLE.value):
+            frappe.throw(_("Only the Added approver can approve"))
+
+        employee_advance_doc.docstatus = 1
+        employee_advance_doc.status = "Pending Approval By Admin"
+        CustomTask.mark_tasks_as_completed(
+            "Employee Advance", doc.name, TaskType.APPROVE_EMPLOYEE_ADVANCE
+        )
+    elif employee_advance_doc.status == "Pending Approval By Admin":
+        if not user_has_role(frappe.session.user, RoleEnum.FINANCE_ROLE.value):
+            frappe.throw(_("Only user having the role of Expense Admin can approve the document"))
+        employee_advance_doc.status = "Pending Approval By HR"
```

Comment:
edit doctype change status to allow on submit

Reply:
Done